### ISSUE TITLE
support for g300s?

### ISSUE THREAD

**Issue by madl-ash** [antonpup/Aurora#1573]
Hello,i just saw this great project and i am wondering if you can support Logitech G300/G300s mouse?
it is not rgb but it has 8 lightning options like this :
R,G,B
255,0,0
0,255,0
0,0,255
255,255,0
255,0,255
0,255,255
255,255,255
0,0,0
i can confirm these work with logitechSDK.
also,people made it to look like rgb by rapidly changing colors with specific timing.

**Comment by Mikowmer** [comment]
I'd like to second this. I also own a g300s.

Currently, I'm assuming that whatever Aurora is sending to the g300s is getting approximated to one of those 8 colour options. Currently, the only work around I can see is to set the Mouse/Headset marker to one of those eight colours, and for every single layer that might have an effect, set up an additional layer with the specific effect and colour. There can be at most only seven (eight if you include 0,0,0) colour gradient options, and a script could probably be written rather easily to utilise that.

In addition, if you cycle through the macro pre-sets using a button on the mouse, it reverts to LGS control, and the colour only reverts back to Aurora's control when the colour on it changes.

Unfortunately, this does mean that your use of other peripherals like headsets and RGB desktop lights are not going to be as flexible as most people would probably like. Also, because it is simpler, it may ironically be more difficult to add support specifically for this mouse.

Admittedly, this is only one person's view on things who doesn't actually know what any of the code looks like, let alone having much coding experience. I know enough to get myself into trouble, and not always enough to get myself out.

**Comment by madl-ash** [comment]
no,you have to set values to either 0 or 255 to work.otherwise it accepts the parameters and wont change the led color.
i'll check the code to see if i can add support or not

**Comment by Mikowmer** [comment]
Erm, not true. I try adjusting the colour directly in aurora, and the led changes at around 120-140 (and seems to be different in Red, Green, Blue, and Opacity). It's rounding either up to full, or down to nothing, probably within the LGS wrapper or within the mouse itself.

Also, another odd effect, at around 120-140 opacity, and 255 for all of RGB, the mouse turns blue, while above is white and below is nothing.

**Comment by madl-ash** [comment]
do you test this with wrapper itself or aurora?
i cant change colors at all in aurora
also you may be right.i used a very old sdk but i  think it must change for values 128+

**Comment by Mikowmer** [comment]
Within Aurora, using latest drivers for LGS, and latest version of Aurora (0.6.4b)

**Comment by madl-ash** [comment]
i can't get it to work in aurora.it wont change colors.what mouse did you choose in settings?

**Comment by Mikowmer** [comment]
Generic Peripheral.

**Comment by madl-ash** [comment]
latest LGS and aurora 0.6.4b and i can't cet my mouse colors to change in anyway.i can change colors using SDK directly.can you tell me step by step how you change colors?so i can look into it

**Comment by Mikowmer** [comment]
Sure. Give me a bit and I'll even post screenshots.

**Comment by Mikowmer** [comment]
Okay. General Settings are like this:
https://imgur.com/vNumE9n
Mostly still set on defaults.
Devices and Wrappers like this:
https://imgur.com/ntUQlSl

On the actual set-up, the Mouse Layer should look like this:
https://imgur.com/ZMFSFaE
Note that the only affected key is the generic Mouse/Headset block. It's a simple Solid Color Layer set at a higher priority than everything else. (Note: MUST BE AT THE TOP OF THE LAYER STACK if you have other layers and you want this colour to stay constant)

To change the colours, simply change the colour of the solid colour layer:
https://imgur.com/w5t4iCf

**Comment by madl-ash** [comment]
SDK is perfectly working but aurora wont change colors.ill look into the code.
btw i checked the SDK documentations.
about g300/g300s:
Colors Supports red on/off, green on/off, blue on/off, or a combination of the three. When calling the SDK’s LogiLedSetLighting function, if the percentage given is below 50, the color will be off, and when above 50, the color will be on.

**Comment by Mikowmer** [comment]
Have you tried completely closing Aurora (Going into task manager and shutting off every instance) and restarting it?

Also, I suspect that the percentages are an integer value as well. 130 is 50.8% of 256, after all. And it appears that 130-131 is the threshold for the colour change.

**Comment by madl-ash** [comment]
yes i did try that.
i also tried patching the logitech SDK with aurora wrapper or re installing LGS and nothing worked.

also about the integers you are right.SDK works with percentage.so 50% and more can change colors
i said 128 because i rounded it to 50% but yea it is accually 129 or 130.also the highest number is 255 not 256

**Comment by Mikowmer** [comment]
Hmm... I've got a pretty clean install at the moment. Only have LGS and Razer Synapse and installed Aurora afterwards... I know this might sound extreme, but have you tried uninstalling both Aurora and LGS and starting from scratch? I apologize if it sounds condescending, it's just that I literally installed stuff and it worked from the get-go.

**Comment by madl-ash** [comment]
so  i uninstalled both and reinstalled again .the problem exists :(

**Comment by madl-ash** [comment]
> 
> Also, another odd effect, at around 120-140 opacity, and 255 for all of RGB, the mouse turns blue, while above is white and below is nothing.

it has something to do with this : https://github.com/antonpup/Aurora/blob/1e635699d55ec112e8f5ff692dfa8d6fd3e83100/Project-Aurora/Project-Aurora/Devices/Logitech/LogitechDevice.cs#L211

**Comment by Mikowmer** [comment]
> so  i uninstalled both and reinstalled again .the problem exists :(

… I give up. It must be something to do with my system that I don't know about that is allowing this to work for me and not for you.



> Also, another odd effect, at around 120-140 opacity, and 255 for all of RGB, the mouse turns blue, while above is white and below is nothing.
> 
> it has something to do with this : 
>   
>     
>       Aurora/Project-Aurora/Project-Aurora/Devices/Logitech/LogitechDevice.cs
>     
>     
>          Line 211
>       in
>       1e63569
>     
>     
>     
>     
> 
>         
>           
>            double alpha_amt = (default_color.A / 255.0);

Still doesn't explain anything, as all values should change at the same rate.
*checks stuff* Ah, that's what's doing it. The odd bits are due to my layer set up. I've got a low level solid color layer on all of my keys to add a bit of low light ambience. I remove that, and the blue bit disappears, goes straight from white to nothing.

And in case that was affecting things, I tried adjusting colours without the base layer. Still works fine.

**Comment by madl-ash** [comment]
okay maybe something is wrong with my system.i'll test it on another computer i guess.and damn...visual studio has problem installing c++ support so it will have to wait sometime before i compile but i guess i can check the code at least

**Comment by Mikowmer** [comment]
Hmm... Reinstall your C++ and your .Net Redistributables. I have a gut feeling that the problem might be there.

**Comment by madl-ash** [comment]
OMFG found the problem
aurora was installed in another drive that was not system drive and was on another hdd
it is working now.
so i guess i reached my goal.it basically does support g300s
what else can we do about it?

**Comment by Mikowmer** [comment]
Might be able to add full support with a mouse image for it, rather than a generic square, and program that image to display the light as one of the eight colours allowed by the mouse, but that's all I can think of for the moment. Beyond that I don't think there would be any more that needs doing.

**Comment by madl-ash** [comment]
About full RGB on the mouse... well I guess this "rounding numbers up" is hard coded in the mouse firmware, so I'm afraid it is nearly impossible...maybe someone can hack the mouse itself as I checked the mouse board and this is a RGB led but I guess it is not digital and there is probably no Analog digital converter for led on the board.
About mouse image i guess it can be implemented.I'm not good with pictures but i can make a image that can replace it for now until someone replaces a good image for it.
For the color selection,i heard the devs are replacing the mouse/keyboard into another GUI that can handle everything(light bulbs,led strips,mousepads and ...) so i guess it will be pointless but maybe some devs can give us a green light and i try to implement it.

**Comment by Mikowmer** [comment]
We aren't going to get full RGB on the mouse. I don't think there is anything we could even hack within the LGS SDK to force it into full RGB without breaking something, and I'd rather my mouse remained intact, thank you very much.

Mouse Image and Colour Selection can probably wait for the new GUI then, unless the Devs already have a general blueprint or design document we can base things off of. Who would we message to get read in on this?

At the very least, the fitting to a colour should be relatively straightforward to implement. I think that all that is holding further movement is design of the new stuff.
